src/witness_generator/machines/fixed_lookup_machine.rs

```diff
@@ -76,60 +74,82 @@ impl FixedLookup {
         left: &[Result<AffineExpression, EvalError>],
         right: &SelectedExpressions,
     ) -> EvalResult {
-        let left_key = match left[0].clone() {
-            Ok(v) => match v.constant_value() {
-                Some(v) => Ok(v),
-                None => Err(format!(
-                    "First expression needs to be constant but is not: {}.",
-                    v.format(fixed_data)
-                )),
-            },
-            Err(err) => Err(format!("First expression on the LHS is unknown: {err}")),
-        }?;
-
-        let right_key = right.expressions.first().unwrap();
-        let rhs_row = if let Expression::PolynomialReference(poly) = right_key {
-            // TODO we really need a search index on this.
-            fixed_data.fixed_cols
-                .get(poly.name.as_str())
-                .and_then(|values| values.iter().position(|v| *v == left_key))
-                .ok_or_else(|| {
-                    format!(
-                        "Unable to find matching row on the RHS where the first element is {left_key} - only fixed columns supported there."
-                    )
-                })
-                .map(|i| i as DegreeType)
-        } else {
-            Err("First item on the RHS must be a polynomial reference.".to_string())
-        }?;
+        let columns = right
+            .expressions
+            .iter()
+            .map(|right_key| {
+                if let Expression::PolynomialReference(poly) = right_key {
+                    fixed_data
+                        .fixed_cols
+                        .get(poly.name.as_str())
+                        .ok_or_else(|| {
+                            format!(
+                                "Unable to find a fixed column for the fixed lookup: {}",
+                                poly.name
+                            )
+                        })
+                } else {
+                    Err("RHS must be a polynomial reference.".to_string())
+                }
+            })
+            .collect::<Result<Vec<_>, _>>()?;
 
-        // TODO we only support the following case:
-        // - The first component on the LHS has to be known
-        // - The first component on the RHS has to be a direct fixed column reference
-        // - The first match of those uniquely determines the rest of the RHS.
+        let len = columns[0].len();
 
-        // TODO in the other cases, we could at least return some improved bit constraints.
+        let res = (0..len)
+            // get all lookup rows which match the lhs
+            .filter_map(|row| {
+                left.iter()
+                    .zip(columns.iter())
+                    .map(|(left, right)| {
+                        let right = &right[row];
```

Comment:
I have the feeling that this might be easier if you do `(left - right.into()).solve()`. The problem is that it does not properly distinguish between "invaild" and "underconstrained". Maybe we should change `solve()` to return a specific error if the affine expression is invalid, then we could just distinguish the three cases here.

Reply:
What do you mean by invalid here? Sorry I'm not familiar enough with `solve` and `invalid`. Maybe this can be a different PR?

Reply:
Yes, it can!